cupy/_core/dlpack.pyx

```diff
@@ -19,139 +14,213 @@ import warnings
 import cupy
 import cupy._core.core as core
 
+from .dlpack cimport *
 
-cdef extern from './include/cupy/_dlpack/dlpack.h' nogil:
-    cdef int DLPACK_VERSION
-
-    cdef enum DLDeviceType:
-        kDLCPU
-        kDLCUDA
-        kDLCUDAHost
-        kDLOpenCL
-        kDLVulkan
-        kDLMetal
-        kDLVPI
-        kDLROCM
-        kDLROCMHost
-        kDLExtDev
-        kDLCUDAManaged
-        kDLOneAPI
-        kDLWebGPU
-        kDLHexagon
-
-    ctypedef struct DLDevice:
-        DLDeviceType device_type
-        int32_t device_id
-
-    cdef enum DLDataTypeCode:
-        kDLInt
-        kDLUInt
-        kDLFloat
-        kDLBfloat
-        kDLComplex
-        kDLBool
-
-    ctypedef struct DLDataType:
-        uint8_t code
-        uint8_t bits
-        uint16_t lanes
-
-    ctypedef struct DLTensor:
-        void* data
-        DLDevice device
-        int32_t ndim
-        DLDataType dtype
-        int64_t* shape
-        int64_t* strides
-        uint64_t byte_offset
-
-    ctypedef struct DLManagedTensor:
-        DLTensor dl_tensor
-        void* manager_ctx
-        void (*deleter)(DLManagedTensor*)  # noqa: E211
-
-
-def get_build_version():
-    return str(DLPACK_VERSION)
+
+cdef const char* CAPSULE_NAME = "dltensor"
+cdef const char* CAPSULE_NAME_VER = "dltensor_versioned"
+cdef const char* USED_CAPSULE_NAME = "used_dltensor"
+cdef const char* USED_CAPSULE_NAME_VER = "used_dltensor_versioned"
 
 
 cdef void pycapsule_deleter(object dltensor):
     cdef DLManagedTensor* dlm_tensor
+    cdef DLManagedTensorVersioned *dlm_tensor_ver
+
     # Do not invoke the deleter on a used capsule
-    if cpython.PyCapsule_IsValid(dltensor, 'dltensor'):
-        dlm_tensor = <DLManagedTensor*>cpython.PyCapsule_GetPointer(
-            dltensor, 'dltensor')
+    if cpython.PyCapsule_IsValid(dltensor, CAPSULE_NAME):
+        dlm_tensor = <DLManagedTensor*>(
+            cpython.PyCapsule_GetPointer(dltensor, CAPSULE_NAME))
         dlm_tensor.deleter(dlm_tensor)
+    elif cpython.PyCapsule_IsValid(dltensor, CAPSULE_NAME_VER):
+        dlm_tensor_ver = <DLManagedTensorVersioned*>(
+            cpython.PyCapsule_GetPointer(dltensor, CAPSULE_NAME_VER))
+        dlm_tensor_ver.deleter(dlm_tensor_ver)
```

Comment:
ditto
```suggestion
        if dlm_tensor_ver.deleter:
            dlm_tensor_ver.deleter(dlm_tensor_ver)
```

Reply:
Here, it's actually fine, since it is our capsule and our dltensor.  Kept it for now as well.

But, we did have the same bug in the DLMemory implementation, so fixed it there (forced pushed, I had thought I had pushed, but forgot the applied suggestions and didn't notice it didn't go through).